driftbase/lobbies.py

```diff
@@ -0,0 +1,757 @@
+                lobby["members"] = [member for member in lobby["members"] if member["player_id"] != member_id]
+
+                kicked = len(lobby["members"]) != current_length
+
+                if kicked:
+                    log.info(f"Host player '{player_id}' kicked member player '{member_id}' from lobby '{lobby_id}'")
+
+                    member_lobby_lock.value = None
+                    lobby_lock.lobby = lobby
+
+                    # Notify members and kicked player
+                    _post_lobby_event_to_members(receiving_player_ids, "LobbyMemberKicked", {"lobby_id": lobby_id, "kicked_player_id": member_id, "members": lobby["members"]})
+                else:
+                    log.warning(f"Host player '{player_id}' tried to kick member player '{member_id}' from lobby '{lobby_id}', but '{member_id}' wasn't a member of the lobby")
+
+                    if member_lobby_id == lobby_id:
+                        log.warning(f"Player '{member_id}' is supposed to be in lobby '{lobby_id}' but isn't a member of the lobby")
+                        member_lobby_lock.value = None
+
+# Helpers
+
+def _internal_join_lobby(player_id: int, lobby_id: str) -> dict:
+    player_name: str = g.db.query(CorePlayer.player_name).filter(CorePlayer.player_id == player_id).first().player_name
+
+    with _LockedLobby(_get_lobby_key(lobby_id)) as lobby_lock:
+        lobby = lobby_lock.lobby
+
+        if not lobby:
+            log.warning(f"Player '{player_id}' attempted to join lobby '{lobby_id}' which doesn't exist")
+            raise NotFoundException(f"Lobby {lobby_id} doesn't exist")
+
+        if not _get_lobby_member(lobby, player_id):
+            lobby["members"].append({
+                "player_id": player_id,
+                "player_name": player_name,
+                "team_name": None,
+                "ready": False,
+                "host": False,
+                "join_date": datetime.datetime.utcnow().isoformat(),
+            })
+
+            lobby_lock.lobby = lobby
+
+            log.info(f"Player '{player_id}' joined lobby '{lobby_id}'")
+
+            # Notify members
+            receiving_player_ids = _get_lobby_member_player_ids(lobby, [player_id])
+            _post_lobby_event_to_members(receiving_player_ids, "LobbyMemberJoined", {"lobby_id": lobby_id, "members": lobby["members"]})
+        else:
+            log.info(f"Player '{player_id}' attempted to join lobby '{lobby_id}' while already being a member")
+
+        return _get_personalized_lobby(lobby, player_id)
+
+def _get_personalized_lobby(lobby: dict, player_id: int) -> dict:
+    # Add personalized connection options if the match has started
+    if _lobby_match_initiated(lobby) and lobby.get("connection_string", None):
+        # TODO: Think about where this code should reside since it's match and GameLift specific
+
+        player_lobby = copy.deepcopy(lobby)
+
+        # Default to spectator
+        connection_options = "SpectatorOnly=1"
```

Comment:
We don't actually assume that server or client is UE4 anywhere else

Reply:
Theoretically the arguments *could* be the same, but this should be put on the refactoring list. Not because we might need it, but because it breaks the platform agnosticism of the app.

Reply:
Added an asana defect for this